### PR TITLE
Test the days in a month and year for the Observational Islamic Calendars

### PR DESCRIPTION
**Pull request by sotam1069**
Added tests to calculate the  amount of days in a month using the `month_length` function from the book
Added tests to calculate the amount of days between year `x` and year `y`, using the data given from the book.

### CODE REVIEW

File: components/calendar/src/islamic.rs [hunk 1]
```diff
@@ -1465,4 +1468,42 @@ mod test {
...
+        // -1245 1 1 = -214526 (R.D Date)
+        // 1518 1 1 = 764589 (R.D Date)
+        let x_year = -1245;
+        let y_year = 1518;
+
+        let mut sum_days_in_year = 0;
+        for i in x_year..y_year {
+            let days = IslamicObservational::days_in_provided_year(i);
+            sum_days_in_year += days;
+        }
```
Comment (echeran) [inline components/calendar/src/islamic.rs:1483]: optional (also not sure if this works):
```suggestion
        let sum_days_in_year = x_year..y_year
           .map(IslamicObservational::days_in_provided_year)
           .sum();
```

**Review by echeran — APPROVED** [review]
LGTM

**Comment by echeran** [comment]
Actually, we should annotate  to ignore these tests since they run a long time (> 60 seconds). It's nice if the wall clock time for unit tests is less than a second because unit tests are typically based on specific minimally-sized examples of functionality.

**Review by echeran — APPROVED** [review]

File: components/calendar/src/islamic.rs [hunk 2]
```diff
@@ -1465,4 +1463,41 @@ mod test {
...
+    fn test_days_in_provided_year_observational() {
+        // -1245 1 1 = -214526 (R.D Date)
+        // 1518 1 1 = 764589 (R.D Date)
+        let x_year = -1245;
+        let y_year = 1518;
+
+        let sum_days_in_year: i32 = (x_year..y_year)
+            .map(|year| IslamicObservational::days_in_provided_year(year) as i32)
+            .sum();
+        let expected_number_of_days = 979115; // The number of days between Islamic years -1245 and 1518
```
Comment (sffc) [inline components/calendar/src/islamic.rs:1478]: Nit, here and below: Test the code that generates this number.

**Review by sffc — COMMENTED** [review]
Would like @Manishearth's review on the change of logic in `days_in_provided_year`

File: components/calendar/src/islamic.rs (hunk 1 above)
Comment (Manishearth) [reply components/calendar/src/islamic.rs]: Yeah that should work, may need some tweaking to get references to work

File: components/calendar/src/islamic.rs [hunk 3]
```diff
@@ -1465,4 +1463,41 @@ mod test {
...
+    #[ignore]
+    #[test]
+    fn test_days_in_provided_year_observational() {
+        // -1245 1 1 = -214526 (R.D Date)
+        // 1518 1 1 = 764589 (R.D Date)
+        let x_year = -1245;
+        let y_year = 1518;
+
+        let sum_days_in_year: i32 = (x_year..y_year)
+            .map(|year| IslamicObservational::days_in_provided_year(year) as i32)
```
Comment (Manishearth) [inline components/calendar/src/islamic.rs:1476]: should we also test that the month length is 29 or 30 with some known exceptions?

**Review by Manishearth — APPROVED** [review]

File: components/calendar/src/islamic.rs (hunk 2 above)
Comment (sotam1069) [reply components/calendar/src/islamic.rs:1478]: done.

**Comment by Manishearth** [comment]
yeah the change is basically idempotent from a logic pov

File: components/calendar/src/islamic.rs [hunk 4]
```diff
@@ -1465,4 +1463,49 @@ mod test {
...
+        // -1245 1 1 = -214526 (R.D Date)
+        // 1518 1 1 = 764589 (R.D Date)
+        let x_year = -1245;
+        let y_year = 1518;
+
+        let sum_days_in_year: i64 = (x_year..y_year)
+            .map(|year| IslamicObservational::days_in_provided_year(year) as i64)
+            .sum();
+        let expected_number_of_days = IslamicObservational::fixed_from_islamic(IslamicDateInner(
+            ArithmeticDate::new_from_lunar_ordinals(1518, 1, 1).unwrap(),
```
Comment (sffc) [inline components/calendar/src/islamic.rs:1479]: Nit, here and below: use `x_year` and `y_year` instead of repeating them.

Comment (sotam1069) [reply components/calendar/src/islamic.rs]: done! I added them as consts.

**Review by sffc — DISMISSED** [review]

**Review by echeran — DISMISSED** [review]

**Review by echeran — APPROVED** [review]

**Pull request merged**